File: heavy-metal-bake-sale/python/tests/product_builder.py
from __future__ import annotations

from heavy_metal_bake_sale import Money, Product


class ProductBuilder:
    def __init__(self) -> None:
        self._name = "Brownie"
        self._price = 0.75
        self._purchase_code = "B"
        self._stock = 48

    def with_name(self, name: str) -> ProductBuilder:
        self._name = name
        return self

    def with_price(self, price: float) -> ProductBuilder:
        self._price = price
        return self

    def with_purchase_code(self, code: str) -> ProductBuilder:
        self._purchase_code = code
        return self

    def with_stock(self, stock: int) -> ProductBuilder:
        self._stock = stock
        return self

    def build(self) -> Product:
        return Product(self._name, Money(self._price), self._purchase_code, self._stock)

File: heavy-metal-bake-sale/typescript/tests/ProductBuilder.ts
import { Money } from '../src/Money.js';
import { Product } from '../src/Product.js';

export class ProductBuilder {
  private _name = 'Brownie';
  private _price = 0.75;
  private _purchaseCode = 'B';
  private _stock = 48;

  withName(name: string): this { this._name = name; return this; }
  withPrice(price: number): this { this._price = price; return this; }
  withPurchaseCode(code: string): this { this._purchaseCode = code; return this; }
  withStock(stock: number): this { this._stock = stock; return this; }

  build(): Product {
    return new Product(this._name, new Money(this._price), this._purchaseCode, this._stock);
  }
}
